### ISSUE TITLE
"Attempting to use module .. which is not loaded"

### ISSUE THREAD

**Issue by utdemir** [haskellari/cabal-doctest#69]
I am trying to add doctest support for the `linear-base` library, which uses GHC 9. 

The usual `cabal-doctests` setup return a bunch of errors along the lines of:

```
/home/utdemir/workspace/github.com/tweag/linear-base/src/Foreign/Marshal/Pure.hs:22: failure in expression `import Data.Unrestricted.Linear'
expected: 
 but got: 
          ^
          <interactive>:1:1: error:
              attempting to use module ‘main:Data.Unrestricted.Linear’ (/home/utdemir/workspace/github.com/tweag/linear-base/src/Data/Unrestricted/Linear.hs) which is not loaded
```

However, a command like: `cabal exec -- doctest -package linear-base src/**/*.hs` succeeds.

I found out that the issue happens because of the `-isrc/` flag introduced by `cabal-doctest`. I think the reason is GHCi trying to use the modules introduced by `-isrc/`, rather than the library in `package-db`, which are not loaded yet.

Here is a slightly modified version of `flags ++ pkgs ++ module_sources` returned by `cabal-doctests`:

```haskell
[ "-i",
  "-i/home/utdemir/workspace/github.com/tweag/linear-base/dist-newstyle/build/x86_64-linux/ghc-9.0.1/linear-base-0.1.1/build/autogen",
  "-i/home/utdemir/workspace/github.com/tweag/linear-base/dist-newstyle/build/x86_64-linux/ghc-9.0.1/linear-base-0.1.1/build",
  -- "-i/home/utdemir/workspace/github.com/tweag/linear-base/src",
  "-package-env=-",
  "-hide-all-packages",
  "-no-user-package-db",
  "-package-db=/home/utdemir/.cabal/store/ghc-9.0.1/package.db",
  "-package-db=/home/utdemir/workspace/github.com/tweag/linear-base/dist-newstyle/packagedb/ghc-9.0.1",
  "-package-db=/home/utdemir/workspace/github.com/tweag/linear-base/dist-newstyle/build/x86_64-linux/ghc-9.0.1/linear-base-0.1.1/package.conf.inplace",
  "-optP-include",
  "-optP/home/utdemir/workspace/github.com/tweag/linear-base/dist-newstyle/build/x86_64-linux/ghc-9.0.1/linear-base-0.1.1/build/autogen/cabal_macros.h",
  "-package-id=base-4.15.0.0",
  "-package-id=containers-0.6.4.1",
  "-package-id=ghc-prim-0.7.0",
  "-package-id=hashable-1.3.1.0-a92c75af1fea2a65b2365eb935f4a4802dac5057158f909650e1011dc3a01a23",
  "-package-id=primitive-0.7.1.0-f003beddf0fc9d04aaf34af7ddb2d21771469f6e4225ce5a793266b89bfbb923",
  "-package-id=storable-tuple-0.0.3.3-7f76b1fd43fb107740fd5c1db4bf4e5a25de9b66b13ca4f53845700662363649",
  "-package-id=text-1.2.4.1",
  "-package-id=transformers-0.5.6.2",
  "-package-id=vector-0.12.3.0-8cc976946fcdbc43a65d82e2ca0ef40a7bb90b17e6cc65c288a8b694f5ac3127",
  "-package-id=doctest-0.18.1-ba4233742e9e8602bcf50dd58a39029934715a541fa504ea2e97f9de2502fb2a",
  "-package=linear-base",
  "src/Foreign/Marshal/Pure.hs"
  -- other modules omitted
]
```

The above list of arguments passes the tests, when the `-i` command including the source directory is commented out. When it is uncommented, the doctests fail with the above error. Also keep in mind that I have to change the list to use the module path instead of the name (`src/Foreign/Marshal/Pure.hs` instead of `Foreign.Marshall.Pure`).

To reproduce the issue (apologies that I couldn't find a smaller example):

```
$ # obtain GHC 9.0.1 and cabal-install 3.4.
$ git clone https://github.com/tweag/linear-base; cd linear-base; git checkout 9b477e4
$ cabal test doctests
```

Relevant linear-base issue: https://github.com/tweag/linear-base/pull/317

**Comment by phadej** [comment]
Honestly I don't understand what is the problem here. Probably somehow you are stretching what `cabal-doctest` is capable of handling. Not adding `-isrc` feels to be wrong and would break many existent users.

```
doctest -package linear-base src/**/*.hs
```

Seems wrong. `doctest` interprets sources, why you need `-package linear-base`?

You have you figure that out.

---

That said, I don't use `cabal-doctest` myself anymore, but rather moved to use `cabal-docspec`, which seems to handle `linear-base` just fine after a small patch:

```diff
diff --git a/src/Foreign/Marshal/Pure.hs b/src/Foreign/Marshal/Pure.hs
index a38ee28..6bd49bf 100644
--- a/src/Foreign/Marshal/Pure.hs
+++ b/src/Foreign/Marshal/Pure.hs
@@ -21,6 +21,7 @@
 -- A toy example:
 --
 -- >>> :set -XLinearTypes
+-- >>> import Prelude
 -- >>> import Data.Unrestricted.Linear
 -- >>> import qualified Foreign.Marshal.Pure as Manual
 -- >>> :{
```

```
% cabal build -w ghc-9.0.1
% cabal-docspec -w ghc-9.0.1
...
Total:        41; Tried:   41; Skipped:    0; Success:   41; Errors:    0; Failures    0
Examples:     41; Tried:   41; Skipped:    0; Success:   41; Errors:    0; Failures    0

[   0.75230] peu.completed: OK
```

But `cabal-docspec` doesn't work with `stack` nor Nix(pkgs) based setup, so it's probably not an option for you.

---

From my perspective this is *wontfix*.

**Comment by ulidtko** [comment]
I also got this issue with `tasty-discover` & `stack`...

(Sorry @phadej — this issue is now top hit on DDG for "stack test attempting to use module which is not loaded")

What purged it for me was `git clean -fxd` / `rm -rf .stack-work` & recompile.

**Comment by andreasabel** [comment]
I am departing as maintainer, only do minimal work like dependency bumping when urgent.
The package is up for grabs, @ulidtko @utdemir :
- #79

**Comment by ulidtko** [comment]
... OK, I'm going to chalk this one up to a PEBKAC **user error**, and close with an explanation.
(Despite the OP no longer using this package anyway - https://github.com/tweag/linear-base/pull/361).

### All fixes are belong to `linear-base`
I was able to reproduce several `Module not loaded` errors, following the repro steps provided by @utdemir:
```
$ git clone https://github.com/tweag/linear-base; cd linear-base; git checkout 9b477e4
$ cabal test doctests
```

I was also able to **fix all of them** ✅ by means of making changes to `linear-base`.

I.e. no change to `cabal-doctest` is necessary.

<details><summary> 🟢 The exact patch</summary>

```diff
From: Max Ulidtko <ulidtko@gmail.com>
Date: Fri, 21 Nov 2025 11:10:47 +0100
Subject: [PATCH] Fix doctests; Re: cabal-doctest#69

Patch on top of https://github.com/tweag/linear-base/commit/9b477e495a33f48da76603e70ede643827faa1c3
which fixes `cabal test doctests`.

---
 src/Data/Array/Mutable/Linear.hs          | 2 +-
 src/Data/Unrestricted/Internal/Dupable.hs | 1 +
 src/Foreign/Marshal/Pure.hs               | 5 +++++
 src/Prelude/Linear.hs                     | 3 +++
 src/System/IO/Resource.hs                 | 3 +++
 test/Doctest.hs                           | 4 ++--
 6 files changed, 15 insertions(+), 3 deletions(-)

diff --git a/src/Data/Array/Mutable/Linear.hs b/src/Data/Array/Mutable/Linear.hs
index 5df29e9..9403afc 100644
--- a/src/Data/Array/Mutable/Linear.hs
+++ b/src/Data/Array/Mutable/Linear.hs
@@ -52,4 +52,4 @@ module Data.Array.Mutable.Linear
 where
 
 import Data.Array.Mutable.Linear.Internal
-import Prelude hiding (read, map)
+import Prelude hiding (read, map, (==), ($))
diff --git a/src/Data/Unrestricted/Internal/Dupable.hs b/src/Data/Unrestricted/Internal/Dupable.hs
index 5732ed0..1d02d45 100644
--- a/src/Data/Unrestricted/Internal/Dupable.hs
+++ b/src/Data/Unrestricted/Internal/Dupable.hs
@@ -16,6 +16,7 @@ import GHC.TypeLits
 import Data.Type.Equality
 import Data.V.Linear.Internal.V (V)
 import qualified Data.V.Linear.Internal.V as V
+import qualified Prelude (Either(..))
 
 -- | The laws of @Dupable@ are dual to those of 'Monoid':
 --
diff --git a/src/Foreign/Marshal/Pure.hs b/src/Foreign/Marshal/Pure.hs
index e6db3b7..e91e97f 100644
--- a/src/Foreign/Marshal/Pure.hs
+++ b/src/Foreign/Marshal/Pure.hs
@@ -63,3 +63,8 @@ module Foreign.Marshal.Pure
   ) where
 
 import Foreign.Marshal.Pure.Internal
+
+-- $setup
+-- >>> :load Data.Unrestricted.Linear Foreign.Marshal.Pure
+-- >>> import Prelude.Linear
+
diff --git a/src/Prelude/Linear.hs b/src/Prelude/Linear.hs
index cd9be93..34c6e06 100644
--- a/src/Prelude/Linear.hs
+++ b/src/Prelude/Linear.hs
@@ -146,6 +146,9 @@ import qualified Prelude
 import Prelude.Linear.Internal
 import Data.String
 
+-- $setup
+-- >>> :load Prelude.Linear
+
 -- | Replacement for the flip function with generalized multiplicities.
 flip :: (a %p -> b %q -> c) %r -> b %q -> a %p -> c
 flip f b a = f a b
diff --git a/src/System/IO/Resource.hs b/src/System/IO/Resource.hs
index 0237f2a..567f3c3 100644
--- a/src/System/IO/Resource.hs
+++ b/src/System/IO/Resource.hs
@@ -57,3 +57,6 @@ module System.IO.Resource
 
 import System.IO.Resource.Internal
 import qualified System.IO as System
+
+-- $setup
+-- >>> :load System.IO.Resource
diff --git a/test/Doctest.hs b/test/Doctest.hs
index adba5b0..6e74e1d 100644
--- a/test/Doctest.hs
+++ b/test/Doctest.hs
@@ -4,9 +4,9 @@ import Build_doctests (flags, pkgs, module_sources)
 import Test.DocTest (doctest)
 
 main :: IO ()
-main = doctest args'
+main = doctest args
   where
-    args = flags ++ pkgs ++ module_sources
+    args = flags ++ pkgs ++ module_sources ++ ["-Wno-deprecated-flags"]
     args' = [ "-i",
               "-i/home/utdemir/workspace/github.com/tweag/linear-base/dist-newstyle/build/x86_64-linux/ghc-9.0.1/linear-base-0.1.1/build/autogen",
               "-i/home/utdemir/workspace/github.com/tweag/linear-base/dist-newstyle/build/x86_64-linux/ghc-9.0.1/linear-base-0.1.1/build",
```
</details> 

The patch reverts to the `args = flags ++ pkgs ++ module_sources` provided by `cabal-doctest`; I also added `-Wno-deprecated-flags` to silence the `TypeInType` deprecation warning.

With this patch applied on top of https://github.com/tweag/linear-base/commit/9b477e495a33f48da76603e70ede643827faa1c3, `cabal test doctests` passes:

```
Running 1 test suites...
Test suite doctests: RUNNING...
Examples: 45  Tried: 45  Errors: 0  Failures: 0
Test suite doctests: PASS
```

## Mistake 1: importing the module from a doctest which it covers itself

Several instances of this general mistake appear:
1. module `Foo` has a doctest,
2. that doctest says `import Foo` before the `module Foo` header,
3. this fails with `attempting to use module 'Foo' which is not loaded`.

The error message is spot on, and the failure is absolutely correct and expected. GHCi refuses to `import` a module in the middle of *loading that same module*.

Example:
```
/tmp/linear-base/src/System/IO/Resource.hs:14: failure in expression `import qualified System.IO.Resource as Linear'
expected: 
 but got: 
          ^
          <interactive>:1:1: error:
              attempting to use module ‘main:System.IO.Resource’ (/tmp/linear-base/src/System/IO/Resource.hs) which is not loaded
```

👆 Notice that it is the doctest of `System.IO.Resource` -- and it is doing `import System.IO.Resource`, and this appears before the module is even *declared*; before its `module System.IO.Resource … where` header.

To fix this, add a [doctest `$setup` block][$setup] that pre-loads the module prior to running any of its doctest blocks:
```
-- $setup
-- >>> :load System.IO.Resource
```

[$setup]: https://hackage.haskell.org/package/doctest-0.24.2#setup-code

Also see the following caveat.


## Doctest `$setup` caveat

As documented in [doctest][$setup], the `$setup` stanza will not work if placed in-between the module header (`module ... where`) and import declarations.

This will not work:
```haskell
module System.IO.Resource
  ( -- * The Resource I/O Monad
    -- …
  ) where

-- $setup
-- >>> :load System.IO.Resource

import System.IO.Resource.Internal
import qualified System.IO as System
```

but this will:
```haskell
module System.IO.Resource
  ( -- * The Resource I/O Monad
    -- …
  ) where

import System.IO.Resource.Internal
import qualified System.IO as System

-- $setup
-- >>> :load System.IO.Resource
```


## Mistake 2: attempting to import a non-loaded module

This is a more complicated variation of Mistake 1; same root cause.

```haskell
-- | …
-- >>> :set -XLinearTypes
-- >>> import Data.Unrestricted.Linear
-- >>> import qualified Foreign.Marshal.Pure as Manual
-- >>> :{
--   nothingWith3 :: Manual.Pool %1-> Ur Int
--   nothingWith3 pool = move (Manual.deconstruct (Manual.alloc 3 pool))
-- :}
--
-- >>> unur (Manual.withPool nothingWith3)
-- 3

module Foreign.Marshal.Pure
  ( …
  ) where

import Foreign.Marshal.Pure.Internal
```

Notice we're committing Mistake 1 again (`import Foreign.Marshal.Pure` from its own pre-header doctest) -- but also, the doctest tries to `import Data.Unrestricted.Linear` which is *outside of the dependencies closure* of `Foreign.Marshal.Pure`.

The module does not import `Data.Unrestricted.Linear`; but the doctest does. By that point of module-loading and import-resolution, `Data.Unrestricted.Linear` happens to be "not ready yet"; from GHCi PoV, it's known to exist but `not loaded`.

Fix simply through pre-loading:

```
-- $setup
-- >>> :load Data.Unrestricted.Linear Foreign.Marshal.Pure
-- >>> import Prelude.Linear
```

And of course, the `Int` mentioned in doctest has to come from somewhere, thus `import Prelude.Linear` as well.


## Mistake 3: sloppy namespacing + NoImplicitPrelude

Understandably, (that revision of) `linear-base` has a `Prelude.Linear` and thus has to deal with `NoImplicitPrelude`. This is easy to get wrong.

Best shown in `Data.Array.Mutable.Linear`:

```haskell
{-# LANGUAGE LinearTypes #-}

-- | …
-- >>> :set -XLinearTypes
-- >>> :set -XNoImplicitPrelude
-- >>> import Prelude.Linear
-- >>> import qualified Data.Array.Mutable.Linear as Array
-- >>> :{
--  isFirstZero :: Array.Array Int %1-> Ur Bool
--  isFirstZero arr =
--    Array.get 0 arr
--      & \(Ur val, arr') -> arr' `lseq` Ur (val == 0)
-- :}
--
-- >>> unur $ Array.fromList [0..10] isFirstZero
-- True
-- >>> unur $ Array.fromList [1,2,3] isFirstZero
-- False
module Data.Array.Mutable.Linear
  ( …
  )
where

import Data.Array.Mutable.Linear.Internal
import Prelude hiding (read, map)
```

This module *compiles* alright; however, its doctest fails with:

```
/tmp/linear-base/src/Data/Array/Mutable/Linear.hs:16: failure in expression `:{
  isFirstZero :: Array.Array Int %1-> Ur Bool
  isFirstZero arr =
    Array.get 0 arr
      & \(Ur val, arr') -> arr' `lseq` Ur (val == 0)
:}'
expected: 
 but got: 
          ^
          <interactive>:62:48: error:
              Ambiguous occurrence ‘==’
              It could refer to
                 either ‘Prelude.==’,
                        imported from ‘Prelude’ at /tmp/linear-base/src/Data/Array/Mutable/Linear.hs:55:1-33
                        (and originally defined in ‘GHC.Classes’)
                     or ‘Prelude.Linear.==’,
                        imported from ‘Prelude.Linear’
                        (and originally defined in ‘Data.Ord.Linear.Internal.Eq’
                           at /tmp/linear-base/src/Data/Ord/Linear/Internal/Eq.hs:30:3-28)
```

This error is also correct, and stems from the (module-private) `import Prelude hiding (read, map)` bringing clashing identifiers into doctest's scope, together with its `>>> import Prelude.Linear` and regardless of `>>> :set -XNoImplicitPrelude`.

This Prelude import is not reexported -- and so, one way to fix the doctest is to add `(==)` and `($)` to the hiding list.

------------------
TL;DR: I don't see any fault in `cabal-doctest` code that can be fixed here. Fixes for all the `not loaded` errors can be achieved completely within user-code (`linear-base`) as shown above.
